packages/flutter/lib/src/material/date_picker.dart

```diff
@@ -1848,18 +1885,43 @@ class _CalendarDateRangePickerState extends State<_CalendarDateRangePicker> {
 
     _startDate = widget.initialStartDate;
     _endDate = widget.initialEndDate;
+    _updateLastSelectableDate();
 
     // Calculate the index for the initially displayed month. This is needed to
     // divide the list of months into two `SliverList`s.
     final DateTime initialDate = widget.initialStartDate ?? widget.currentDate;
-    if (!initialDate.isBefore(widget.firstDate) &&
-        !initialDate.isAfter(widget.lastDate)) {
+    if (!initialDate.isBefore(widget.firstDate) && !initialDate.isAfter(widget.lastDate)) {
       _initialMonthIndex = DateUtils.monthDelta(widget.firstDate, initialDate);
     }
 
     _showWeekBottomDivider = _initialMonthIndex != 0;
   }
 
+  void _updateLastSelectableDate() {
+    if (widget.selectableDayPredicate == null) {
+      return;
+    }
+    if (_startDate == null || _endDate != null) {
+      _lastSelectableDate = null;
+      return;
+    }
+
+    DateTime? nextNonSelectableDate;
+    // loop until reaching the last date or the next non-selectable date
+    for (DateTime nextDate = _startDate!.add(const Duration(days: 1));
```

Comment:
This should be changed to allow the use case where a client wants to allow selection over selectable dates.  For example say in the following days 1 to 7 (monday to sunday) the wednesday should not be selectable, there are valid use case to allow the selection of start day being monday to sunday.

Both scenario can be achieved by selectableDayPredicate in and of itself. For example if you do not want a range to go over unselectable dates you can change your selectableDayPredicate to check whether there is a current start date being selected and if so remove all dates after the next unselectable dates.

Reply:
The meaning of a non selectable day is... Not being able to select it, so not being able to include it in a range. You are proposing more a way of including the non selectable date that could not be the start or endDate?

It seems quite counter intuitive in that case, maybe I did not understand 🤔

Reply:
Or maybe we could just provide the day to validate **and** the selected start day in the predicate. That way, users can disable the days after the next non selectable day using on the selectable day predicate!

Reply:
But still, for the input date range picker, including a non selectable day in the selected range seems really weird to me, that feels use-case specific

Reply:
Great use case. How should we display non-selectable days as normal days inside a selected range ? I would say that it's okay as of today, because the range would represent the start-to-the-end range.

Let's take in the following image with the selected range (from day 9 to day 12). Is it okay to still display the range like this even if the day 10 is not selectable ? Or do we want to grey the day number text ?

<img width="371" alt="image" src="https://github.com/flutter/flutter/assets/16419143/0f31ed0a-3edf-4779-aad7-eb9b9403621c">

Reply:
I'd say the unselectable days that are in the selection should have their text look a bit more like when there is no selection (more grey rather than black in your example).

----------

This may is probably out of scope for this PR, but , ultimately I believe some sort of builder should be providable to let the user customize this as it pleases him.

Just an example pseudo code
```
dayCellBuilder: (day, selectedStartDate, selectedEndDate, isSelectable) => DayCell(
  day: day,
  textStyle: buildTextStyle(day, selectedStartDate, selectedEndDate, isSelectable),
  boxDecoration: buildBoxDecoration(day, selectedStartDate, selectedEndDate, isSelectable),
)
```

Reply:
No problem, it could be interesting, but isn't it out of the scope of flutter that provides material designs ? I would love to contribute to provide a way to customize the picker btw (to stop having all those community maintained plugins just because of this customization limitation of the flutter framework).

---

So should I keep the current behavior (a non-selectable day looks like the same as selectable day when inside a selected date range), or should I put the text to the disabled color (grey) ?

Reply:
I'd recommend putting the disabled color by default. 

--------

Concerning the picker customization (which I don't recommend to have in this PR), widgets can have a default material design style while still being able to be customizable. `Text`, `ElevatedButton`, etc, while smaller widgets still can be customized to look different.

Reply:
It looks great and coherent!
<img width="363" alt="image" src="https://github.com/flutter/flutter/assets/16419143/21e32def-46f0-41d6-974f-281a0405e28e">